### ISSUE TITLE
VS Code doesn't find the types/library

### ISSUE THREAD

**Issue by aJonathanSchneider** [wikibus/Alcaeus#5]
I tried using heracles with just the angular quickstart tutorial, and so far Visual Studio doesn't recognize heracles typings unless I add this line to "typings/index.d.ts": 
`/// <reference path="../node_modules/heracles/heracles.d.ts" />`
When running
`typings install npm~heracles --global`
it says there's no entry for heracles.

So far, when trying to install/run heracles just through the steps in the readme, it won't work. Could you update it, so that newcomers have a simpler start?

And thanks a lot for publishing this repository, looks like a good project!

**Comment by tpluscode** [comment]
Hi. I haven't used heracles' typings with VS. Is that VS Code you are using?

Also I'm not sure you should be using `--global` flag. Isn't that for libraries which live in global namespace like jQuery? Please try installing without that flag.

**Comment by aJonathanSchneider** [comment]
Thanks for the quick reply!
The error is pretty much the same without the `--global` flag. If I understand the documentation for typings correctly, the `global` flag is a replacement for `ambient`, but doesn't mean operating system/environment-wide (like the npm -g flag), but rather project-wide (as in the folder where your package.json and index.html are). It might also be that a typings-definition has to be manually added to their registry, at least that's what the error message says:

d:\dev\ws\bizSys\bizsys-prototype>typings install npm~heracles                                                                                                                                     
typings ERR! message Unable to find "heracles" ("npm") in the registry.                                                                                                                            
typings ERR! message We could use your help adding these typings to the registry: https://github.com/typings/registry                                                                              
typings ERR! caused by https://api.typings.org/entries/npm/heracles/versions/latest responded with 404, expected it to equal 200                                                                   
                                                                                                                                                                                                   
typings ERR! cwd d:\dev\ws\bizSys\bizsys-prototype                                                                                                                                                 
typings ERR! system Windows_NT 10.0.14393                                                                                                                                                          
typings ERR! command "C:\\Program Files\\nodejs\\node.exe" "C:\\Users\\John\\AppData\\Roaming\\npm\\node_modules\\typings\\dist\\bin.js" "install" "npm~heracles"                                  
typings ERR! node -v v6.5.0                                                                                                                                                                        
typings ERR! typings -v 1.4.0                                                                                                                                                                      
                                                                                                                                                                                                   
typings ERR! If you need help, you may report this error at:                                                                                                                                       
typings ERR!   <https://github.com/typings/typings/issues> 


It seems as if typings only automatically gets the correct types under certain conditions, one of which would be an index.d.ts file in the library root, or a `types` or `typings` property in package.json. If I rename the heracles.d.ts in npm_modules/heracles to index.d.ts I get errors and warnings saying that that wouldn't be a valid component.

**Comment by tpluscode** [comment]
I think you're right. I will add heracles to typings registry.

Until it's done you can use github directly:

```
typings install heracles=github:wikibus/heracles/heracles.d.ts#v0.1.0 
```

**Comment by aJonathanSchneider** [comment]
I got it to load now, but I had to do some adjustments:

- some library expects an index.d.ts at the root of the package, so I renamed heracles.d.ts to index.d.ts
- inside index.d.ts I prefixed everything with `export`, both in the package and in the typings-folder
- to the config.js generated by jspm I added these lines:
`"heracles": "npm:heracles@0.1.1",`
`'rdf-ext':"npm:rdf-ext@0.3.0",`
because for some reason it was trying to load rdf-ext.js from the main directory

**Comment by tpluscode** [comment]
So you installed from JSPM and not directly from NPM right?

**Comment by aJonathanSchneider** [comment]
I installed it both with npm and jspm, I guess the IDE is looking it up for inspection in node_modules, while config.js is set to load it from jspm. I haven't yet managed to load it only through npm

**Comment by tpluscode** [comment]
I tried NPM install for lib and typings but the latter fail with some circular dependency, which I don't understand. I will probably push to the registry instead. At least for the time being.

Could you PR the changes you made? I'll test them out and get my registry PR merged

**Comment by aJonathanSchneider** [comment]
added a pull request with my experiments to recompile the library. The project I'm working on takes the unmodified version that you get with `npm/jspm install`, here's a link: https://github.com/aJonathanSchneider/bizsys-prototype (use `npm start`) . Modified versions of index.d.ts have been added as well

**Comment by tpluscode** [comment]
Okay, so I merged your changes, slightly modified. See if you're mostly happy with them.

Now, I'd like to merge typings/registry#828 but the way you created the `interfaces.ts` file doesn't work. Typings expects `interfaces.d.ts`. As soon as I get this resolved you should be able to install typings

**Comment by aJonathanSchneider** [comment]
Wow, that was quick! I was sick the other week and didn't have a look at
the changes yet. I think it should end in .d.ts if it's supposed to be a
module. I will look into the changes and give you feedback once I've looked
into them

**Comment by tpluscode** [comment]
All right! The typings are live. You should now be able to install by

```
typings i heracles --save
```

**Comment by aJonathanSchneider** [comment]
Hey, I checked out the changes, and now it actually loads the typings even without explicit typings-installation, which is pretty nice. I tried using it through npm only, and now the result is that when executing the tsc-command it tries to interpret the library. While doing so, it doesn't find a lot of the modules and throws some other errors. Interestingly, only if I actually include code referencing heracles in the project. I made two commits to the project where I'm using heracles, directly comparing it:
[crashing](https://github.com/aJonathanSchneider/bizsys-prototype/commit/cbf83e46d422a738d9c9cd873c9fb285d854a693)
[starting, but without really using heracles](https://github.com/aJonathanSchneider/bizsys-prototype/commit/a117580cab84ee54762ecb424b9271f6e7920bdd)

Currently that's the stack trace I'm getting:
`app/app.component.ts(2,10): error TS2305: Module ''heracles'' has no exported member 'Hydra'.                                                                                                    
node_modules/heracles/src/ApiDocumentation.ts(7,40): error TS2307: Cannot find module 'core-decorators/lib/nonenumerable'.                                                                       
node_modules/heracles/src/ApiDocumentation.ts(33,16): error TS2322: Type '{}' is not assignable to type 'IHeracles'.                                                                             
node_modules/heracles/src/FetchUtil.ts(3,21): error TS2307: Cannot find module 'li'.                                                                                                             
node_modules/heracles/src/FetchUtil.ts(5,34): error TS2307: Cannot find module 'jsonld'.                                                                                                         
node_modules/heracles/src/FetchUtil.ts(7,30): error TS2307: Cannot find module 'jsonld'.                                                                                                         
node_modules/heracles/src/FetchUtil.ts(8,23): error TS2307: Cannot find module 'rdf-ext'.                                                                                                        
node_modules/heracles/src/FetchUtil.ts(9,26): error TS2307: Cannot find module 'rdf-formats-common'.                                                                                             
node_modules/heracles/src/FetchUtil.ts(10,35): error TS2307: Cannot find module 'rdf-serializer-jsonld'.                                                                                         
node_modules/heracles/src/FetchUtil.ts(32,23): error TS2339: Property 'fetch' does not exist on type 'Window'.                                                                                   
node_modules/heracles/src/FetchUtil.ts(38,24): error TS2304: Cannot find name 'Response'.                                                                                                        
node_modules/heracles/src/FetchUtil.ts(49,23): error TS2339: Property 'fetch' does not exist on type 'Window'.                                                                                   
node_modules/heracles/src/FetchUtil.ts(57,24): error TS2304: Cannot find name 'Response'.                                                                                                        
node_modules/heracles/src/FetchUtil.ts(67,35): error TS2304: Cannot find name 'Response'.                                                                                                        
node_modules/heracles/src/FetchUtil.ts(75,34): error TS2304: Cannot find name 'Response'.                                                                                                        
node_modules/heracles/src/FetchUtil.ts(99,26): error TS2304: Cannot find name 'Response'.                                                                                                        
node_modules/heracles/src/FetchUtil.ts(110,31): error TS2304: Cannot find name 'Response'.                                                                                                       
node_modules/heracles/src/Resources.ts(3,34): error TS2307: Cannot find module 'jsonld'.                                                                                                         
node_modules/heracles/src/Resources.ts(4,32): error TS2307: Cannot find module 'core-decorators/lib/nonenumerable'.                                                                              
node_modules/heracles/src/Resources.ts(68,16): error TS2322: Type '{}' is not assignable to type 'IApiDocumentation'.                                                                            
  Property 'classes' is missing in type '{}'.                                                                                                                                                    
node_modules/heracles/src/Resources.ts(77,16): error TS2322: Type '{}' is not assignable to type 'IIncomingLink[]'.                                                                              
  Property 'length' is missing in type '{}'.                                                                                                                                                     
node_modules/heracles/src/Resources.ts(98,54): error TS2345: Argument of type '{}' is not assignable to parameter of type 'IHeracles'.                                                           
  Property 'resourceFactory' is missing in type '{}'.                                                                                                                                            
node_modules/heracles/src/Resources.ts(145,16): error TS2322: Type '{}' is not assignable to type 'ISupportedOperation'.                                                                         
  Property 'method' is missing in type '{}'.                                                                                                                                                     
node_modules/heracles/src/Resources.ts(150,16): error TS2322: Type '{}' is not assignable to type 'IResource'.                                                                                   
  Property 'id' is missing in type '{}'.                                                                                                                                                         
node_modules/heracles/src/Resources.ts(155,16): error TS2322: Type '{}' is not assignable to type 'IHeracles'.                                                                                   
node_modules/heracles/src/heracles.ts(1,1): error TS6053: File 'D:/dev/ws/bizSys/bizsys-prototype/node_modules/heracles/typings/index.d.ts' not found.     `
